In [37]:
import os

import pandas as pd

import common

# Assign notebook and folder names
notebook_name = '01_subset_metadata'
figure_folder = os.path.join(common.FIGURE_FOLDER, notebook_name)
data_folder = os.path.join(common.DATA_FOLDER, notebook_name)
print('Figure folder:', figure_folder)
print('Data folder:', data_folder)

# Make the folders
! mkdir -p $figure_folder
! mkdir -p $data_folder

Figure folder: ../figures/01_subset_metadata
Data folder: ../data/01_subset_metadata


First read in the table that contains all the metadata. Print the length of the table (to know how many cells we are dealing with) and look at the first 5 rows.

In [53]:
metadata = pd.read_table("../data/00_original/E-MTAB-5061.sdrf.txt", index_col=0)
print(len(metadata))
metadata.head(5)

3514


Comment[ENA_SAMPLE] Comment[BioSD_SAMPLE]  \
Source Name                                             
AZ_A1                ERS1348470          SAMEA4437021   
AZ_A10               ERS1348471          SAMEA4437022   
AZ_A11               ERS1348472          SAMEA4437023   
AZ_A12               ERS1348473          SAMEA4437024   
AZ_A2                ERS1348474          SAMEA4437025   

            Characteristics[organism] Characteristics[organism part]  \
Source Name                                                            
AZ_A1                    Homo sapiens                       pancreas   
AZ_A10                   Homo sapiens                       pancreas   
AZ_A11                   Homo sapiens                       pancreas   
AZ_A12                   Homo sapiens                       pancreas   
AZ_A2                    Homo sapiens                       pancreas   

            Characteristics[individual]  \
Source Name                               
AZ_A1                                AZ   
AZ_A10                               AZ   
AZ_A11                               AZ   
AZ_A12                               AZ   
AZ_A2                                AZ   

            Characteristics[single cell well quality]  \
Source Name                                             
AZ_A1                                low quality cell   
AZ_A10                                             OK   
AZ_A11                                             OK   
AZ_A12                                             OK   
AZ_A2                                              OK   

            Characteristics[cell type] Characteristics[disease]  \
Source Name                                                       
AZ_A1                   not applicable                   normal   
AZ_A10                      delta cell                   normal   
AZ_A11                      alpha cell                   normal   
AZ_A12                      delta cell                   normal   
AZ_A2                       gamma cell                   normal   

            Characteristics[sex]  Characteristics[age]  \
Source Name                                              
AZ_A1                       male                    43   
AZ_A10                      male                    43   
AZ_A11                      male                    43   
AZ_A12                      male                    43   
AZ_A2                       male                    43   

                      ...           Comment[ENA_EXPERIMENT]        Scan Name  \
Source Name           ...                                                      
AZ_A1                 ...                        ERX1700346   AZ_A1.fastq.gz   
AZ_A10                ...                        ERX1700347  AZ_A10.fastq.gz   
AZ_A11                ...                        ERX1700348  AZ_A11.fastq.gz   
AZ_A12                ...                        ERX1700349  AZ_A12.fastq.gz   
AZ_A2                 ...                        ERX1700350   AZ_A2.fastq.gz   

            Comment[SUBMITTED_FILE_NAME]  Comment[ENA_RUN]  \
Source Name                                                  
AZ_A1                     AZ_A1.fastq.gz        ERR1630013   
AZ_A10                   AZ_A10.fastq.gz        ERR1630014   
AZ_A11                   AZ_A11.fastq.gz        ERR1630015   
AZ_A12                   AZ_A12.fastq.gz        ERR1630016   
AZ_A2                     AZ_A2.fastq.gz        ERR1630017   

                                            Comment[FASTQ_URI] Protocol REF.3  \
Source Name                                                                     
AZ_A1        ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR163/003/...   P-MTAB-51966   
AZ_A10       ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR163/004/...   P-MTAB-51966   
AZ_A11       ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR163/005/...   P-MTAB-51966   
AZ_A12       ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR163/006/...   P-MTAB-51966   
AZ_A2        ftp://ftp.sra.ebi.ac.uk/vol1/fast

We are only going to keep a small number of cells to analyze on our machines locally. Let's define a few cell populations to keep for the analysis. 


I'm going to take a look at the data to see how many cells were collected per patient, per cell type. First, groupby will collect by patient age and cell type. Then I will call a specific column to count (in this case 'Comment[ENA_SAMPLE]') which will give me a count of the number of cells in each age and cell type

In [68]:
metadata.groupby(by = ['Characteristics[age]','Characteristics[cell type]'])['Comment[ENA_SAMPLE]'].count()



Characteristics[age]  Characteristics[cell type] 
22                    PSC cell                         6
                      acinar cell                      3
                      alpha cell                     136
                      beta cell                       34
                      co-expression cell               6
                      delta cell                       7
                      ductal cell                      8
                      endothelial cell                 1
                      gamma cell                       2
                      not applicable                 180
23                    PSC cell                        10
                      acinar cell                      3
                      alpha cell                      92
                      beta cell                       35
                      co-expression cell               6
                      delta cell                      12
                      ductal cell     

We are going to focus on the endocrine cells, so I am going to subset the metadata only for endocrine

In [69]:
cell_types_to_keep = ['alpha cell','beta cell','gamma cell','delta cell',
                      'unclassified endocrine cell','epsilon cell']


This is a subset of endocrine cells (without the alpha cells) so I am going to call this subset endocrine only. With .loc I am pointing to a COLUMN in medata that is called 'Characteristics[cell type]'. Then I am keeping all the rows that contain a value in that column that is in my list defined above. This cuts it down to 629 cells.

In [70]:
endocrine_only = metadata.loc[metadata['Characteristics[cell type]'].isin(cell_types_to_keep)]
print(len(endocrine_only))
endocrine_only.head()

1515


Comment[ENA_SAMPLE] Comment[BioSD_SAMPLE]  \
Source Name                                             
AZ_A10               ERS1348471          SAMEA4437022   
AZ_A11               ERS1348472          SAMEA4437023   
AZ_A12               ERS1348473          SAMEA4437024   
AZ_A2                ERS1348474          SAMEA4437025   
AZ_A6                ERS1348478          SAMEA4437029   

            Characteristics[organism] Characteristics[organism part]  \
Source Name                                                            
AZ_A10                   Homo sapiens                       pancreas   
AZ_A11                   Homo sapiens                       pancreas   
AZ_A12                   Homo sapiens                       pancreas   
AZ_A2                    Homo sapiens                       pancreas   
AZ_A6                    Homo sapiens                       pancreas   

            Characteristics[individual]  \
Source Name                               
AZ_A10                               AZ   
AZ_A11                               AZ   
AZ_A12                               AZ   
AZ_A2                                AZ   
AZ_A6                                AZ   

            Characteristics[single cell well quality]  \
Source Name                                             
AZ_A10                                             OK   
AZ_A11                                             OK   
AZ_A12                                             OK   
AZ_A2                                              OK   
AZ_A6                                              OK   

            Characteristics[cell type] Characteristics[disease]  \
Source Name                                                       
AZ_A10                      delta cell                   normal   
AZ_A11                      alpha cell                   normal   
AZ_A12                      delta cell                   normal   
AZ_A2                       gamma cell                   normal   
AZ_A6                       alpha cell                   normal   

            Characteristics[sex]  Characteristics[age]  \
Source Name                                              
AZ_A10                      male                    43   
AZ_A11                      male                    43   
AZ_A12                      male                    43   
AZ_A2                       male                    43   
AZ_A6                       male                    43   

                      ...           Comment[ENA_EXPERIMENT]        Scan Name  \
Source Name           ...                                                      
AZ_A10                ...                        ERX1700347  AZ_A10.fastq.gz   
AZ_A11                ...                        ERX1700348  AZ_A11.fastq.gz   
AZ_A12                ...                        ERX1700349  AZ_A12.fastq.gz   
AZ_A2                 ...                        ERX1700350   AZ_A2.fastq.gz   
AZ_A6                 ...                        ERX1700354   AZ_A6.fastq.gz   

            Comment[SUBMITTED_FILE_NAME]  Comment[ENA_RUN]  \
Source Name                                                  
AZ_A10                   AZ_A10.fastq.gz        ERR1630014   
AZ_A11                   AZ_A11.fastq.gz        ERR1630015   
AZ_A12                   AZ_A12.fastq.gz        ERR1630016   
AZ_A2                     AZ_A2.fastq.gz        ERR1630017   
AZ_A6                     AZ_A6.fastq.gz        ERR1630021   

                                            Comment[FASTQ_URI] Protocol REF.3  \
Source Name                                                                     
AZ_A10       ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR163/004/...   P-MTAB-51966   
AZ_A11       ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR163/005/...   P-MTAB-51966   
AZ_A12       ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR163/006/...   P-MTAB-51966   
AZ_A2        ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR163/007/...   P-MTAB-51966   
AZ_A6        ftp://ftp.sra.ebi.ac.uk/vol1/fast

That's still too many cells. So let's only keep cells from 6 people, 3 healthy and 3 T2D

How many cells came from each person in this group? In this data, the unique people are marked with a different age. So I am going to count the number of time each age value appears in the age column with value_counts()

This tells me that 170 cells were collected from the donor that was 22 years old and so on... 

In [71]:
endocrine_only['Characteristics[age]'].value_counts()

57    234
55    230
25    221
22    179
23    178
37    150
52    119
48     80
27     65
43     59
Name: Characteristics[age], dtype: int64

I am going to pick 6 people to keep. 

In [94]:
ages_to_keep = ['43','48','47','57','52','55']

endocrine_only_6_samples = endocrine_only.loc[endocrine_only['Characteristics[age]'].isin(ages_to_keep)]
print(len(endocrine_only_6_samples))
endocrine_only_6_samples.head()

722


Comment[ENA_SAMPLE] Comment[BioSD_SAMPLE]  \
Source Name                                             
AZ_A10               ERS1348471          SAMEA4437022   
AZ_A11               ERS1348472          SAMEA4437023   
AZ_A12               ERS1348473          SAMEA4437024   
AZ_A2                ERS1348474          SAMEA4437025   
AZ_A6                ERS1348478          SAMEA4437029   

            Characteristics[organism] Characteristics[organism part]  \
Source Name                                                            
AZ_A10                   Homo sapiens                       pancreas   
AZ_A11                   Homo sapiens                       pancreas   
AZ_A12                   Homo sapiens                       pancreas   
AZ_A2                    Homo sapiens                       pancreas   
AZ_A6                    Homo sapiens                       pancreas   

            Characteristics[individual]  \
Source Name                               
AZ_A10                               AZ   
AZ_A11                               AZ   
AZ_A12                               AZ   
AZ_A2                                AZ   
AZ_A6                                AZ   

            Characteristics[single cell well quality]  \
Source Name                                             
AZ_A10                                             OK   
AZ_A11                                             OK   
AZ_A12                                             OK   
AZ_A2                                              OK   
AZ_A6                                              OK   

            Characteristics[cell type] Characteristics[disease]  \
Source Name                                                       
AZ_A10                      delta cell                   normal   
AZ_A11                      alpha cell                   normal   
AZ_A12                      delta cell                   normal   
AZ_A2                       gamma cell                   normal   
AZ_A6                       alpha cell                   normal   

            Characteristics[sex]  Characteristics[age]  \
Source Name                                              
AZ_A10                      male                    43   
AZ_A11                      male                    43   
AZ_A12                      male                    43   
AZ_A2                       male                    43   
AZ_A6                       male                    43   

                      ...           Comment[ENA_EXPERIMENT]        Scan Name  \
Source Name           ...                                                      
AZ_A10                ...                        ERX1700347  AZ_A10.fastq.gz   
AZ_A11                ...                        ERX1700348  AZ_A11.fastq.gz   
AZ_A12                ...                        ERX1700349  AZ_A12.fastq.gz   
AZ_A2                 ...                        ERX1700350   AZ_A2.fastq.gz   
AZ_A6                 ...                        ERX1700354   AZ_A6.fastq.gz   

            Comment[SUBMITTED_FILE_NAME]  Comment[ENA_RUN]  \
Source Name                                                  
AZ_A10                   AZ_A10.fastq.gz        ERR1630014   
AZ_A11                   AZ_A11.fastq.gz        ERR1630015   
AZ_A12                   AZ_A12.fastq.gz        ERR1630016   
AZ_A2                     AZ_A2.fastq.gz        ERR1630017   
AZ_A6                     AZ_A6.fastq.gz        ERR1630021   

                                            Comment[FASTQ_URI] Protocol REF.3  \
Source Name                                                                     
AZ_A10       ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR163/004/...   P-MTAB-51966   
AZ_A11       ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR163/005/...   P-MTAB-51966   
AZ_A12       ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR163/006/...   P-MTAB-51966   
AZ_A2        ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR163/007/...   P-MTAB-51966   
AZ_A6        ftp://ftp.sra.ebi.ac.uk/vol1/fast

That's still too many cells, so I am only going to keep 50 alpha cells per person in addition to all the rest of the enodcrine cells. 

There is a lot going on in this loop. I will break down each step. 

I am going to loop over the ages to keep because for each person, I need to grab 50 alpha cells. 

The first row grabs all the rows in the dataframe that come from one person. 

Then I grab only the rows that are alpha cells. 

I define rows_to_keep as the first 50 alpha cells (head(50))

I make a new dataframe to be everything from the old one EXCEPT the alpha cells.

Then I make a new dataframe (df_final) with only 50 alpha rows, plus the dataframe that has no alpha cells. 

I print the lengths of each to check that I am really adding 50 for each one. 

Then at the end I am saving that dataframe into a list of dataframes that I opened in the beginning (called all_dfs). After the loop finishes, I am merging all of the dataframes back into one with pd.concat(all_dfs) leaving us with a metadata_subset. 

In [95]:
all_dfs = []

for i in ages_to_keep:
    df = endocrine_only_6_samples.loc[endocrine_only_6_samples['Characteristics[age]'] == int(i)]
    alpha_rows = df.loc[df['Characteristics[cell type]'] == 'alpha cell']
    alpha_rows_to_keep = alpha_rows.head(50)
    df_new = df.loc[df['Characteristics[cell type]'] != 'alpha cell']
    print(len(df_new))
    df_final = pd.concat([df_new, alpha_rows_to_keep])
    print(len(df_final))
    all_dfs.append(df_final)

metadata_subset = pd.concat(all_dfs)

31
59
54
80
0
0
93
143
32
82
134
184


In [96]:
print(len(metadata_subset))
metadata_subset.head()

548


Comment[ENA_SAMPLE] Comment[BioSD_SAMPLE]  \
Source Name                                             
AZ_A10               ERS1348471          SAMEA4437022   
AZ_A12               ERS1348473          SAMEA4437024   
AZ_A2                ERS1348474          SAMEA4437025   
AZ_A8                ERS1348480          SAMEA4437031   
AZ_A9                ERS1348481          SAMEA4437032   

            Characteristics[organism] Characteristics[organism part]  \
Source Name                                                            
AZ_A10                   Homo sapiens                       pancreas   
AZ_A12                   Homo sapiens                       pancreas   
AZ_A2                    Homo sapiens                       pancreas   
AZ_A8                    Homo sapiens                       pancreas   
AZ_A9                    Homo sapiens                       pancreas   

            Characteristics[individual]  \
Source Name                               
AZ_A10                               AZ   
AZ_A12                               AZ   
AZ_A2                                AZ   
AZ_A8                                AZ   
AZ_A9                                AZ   

            Characteristics[single cell well quality]  \
Source Name                                             
AZ_A10                                             OK   
AZ_A12                                             OK   
AZ_A2                                              OK   
AZ_A8                                              OK   
AZ_A9                                              OK   

              Characteristics[cell type] Characteristics[disease]  \
Source Name                                                         
AZ_A10                        delta cell                   normal   
AZ_A12                        delta cell                   normal   
AZ_A2                         gamma cell                   normal   
AZ_A8                          beta cell                   normal   
AZ_A9        unclassified endocrine cell                   normal   

            Characteristics[sex]  Characteristics[age]  \
Source Name                                              
AZ_A10                      male                    43   
AZ_A12                      male                    43   
AZ_A2                       male                    43   
AZ_A8                       male                    43   
AZ_A9                       male                    43   

                        ...              Comment[ENA_EXPERIMENT]  \
Source Name             ...                                        
AZ_A10                  ...                           ERX1700347   
AZ_A12                  ...                           ERX1700349   
AZ_A2                   ...                           ERX1700350   
AZ_A8                   ...                           ERX1700356   
AZ_A9                   ...                           ERX1700357   

                   Scan Name Comment[SUBMITTED_FILE_NAME]  Comment[ENA_RUN]  \
Source Name                                                                   
AZ_A10       AZ_A10.fastq.gz              AZ_A10.fastq.gz        ERR1630014   
AZ_A12       AZ_A12.fastq.gz              AZ_A12.fastq.gz        ERR1630016   
AZ_A2         AZ_A2.fastq.gz               AZ_A2.fastq.gz        ERR1630017   
AZ_A8         AZ_A8.fastq.gz               AZ_A8.fastq.gz        ERR1630023   
AZ_A9         AZ_A9.fastq.gz               AZ_A9.fastq.gz        ERR1630024   

                                            Comment[FASTQ_URI] Protocol REF.3  \
Source Name                                                                     
AZ_A10       ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR163/004/...   P-MTAB-51966   
AZ_A12       ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR163/006/...   P-MTAB-51966   
AZ_A2        ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR163/007/...   P-MTAB-51966   
AZ_A8        ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR163/003/...   P-MTAB-51966 

548 cells! Great, we're happy with that number, so let's save the subset file as a csv)

In [99]:
metadata_subset.to_csv(data_folder+"/metadata_subset_548_cells.csv")